contracts/vesting/src/storage.rs

```diff
@@ -0,0 +1,189 @@
+use curve::Curve;
+use soroban_sdk::{
+    contracttype, log, panic_with_error, Address, ConversionError, Env, String, TryFromVal, Val,
+    Vec,
+};
+
+use crate::error::ContractError;
+
+impl TryFromVal<Env, DataKey> for Val {
+    type Error = ConversionError;
+
+    fn try_from_val(_env: &Env, v: &DataKey) -> Result<Self, Self::Error> {
+        Ok((*v as u32).into())
+    }
+}
+
+#[derive(Clone, Copy)]
+#[repr(u32)]
+pub enum DataKey {
+    Admin = 1,
+    Config = 2,
+    Minter = 3,
+    Whitelist = 4,
+    VestingTokenInfo = 5,
+    MaxVestingComplexity = 6,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VestingTokenInfo {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u32,
+    pub address: Address,
+    pub total_supply: i128,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VestingBalance {
+    pub address: Address,
+    pub balance: i128,
+    pub curve: Curve,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MinterInfo {
+    pub address: Address,
+    pub capacity: Curve,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VestingInfo {
+    pub amount: i128,
+    pub curve: Curve,
+}
+
+pub fn save_admin(env: &Env, admin: &Address) {
+    env.storage().persistent().set(&DataKey::Admin, admin);
+}
+
+pub fn get_admin(env: &Env) -> Address {
+    env.storage()
+        .persistent()
+        .get(&DataKey::Admin)
+        .unwrap_or_else(|| {
+            log!(&env, "Vesting: Get admin: Critical error - No admin found");
+            panic_with_error!(env, ContractError::NoAdminFound);
+        })
+}
+
+pub fn save_balance(env: &Env, address: &Address, balance: i128) {
+    env.storage().persistent().set(address, &balance);
```

Comment:
Is this needed? Usually the internal types doesn't need references.

```suggestion
pub fn save_balance(env: &Env, address: &Address, balance: i128) {
    env.storage().persistent().set(address, balance);
```

Reply:
thought that it should work, but the compiler does not allow it:

```Rust
contracts/vesting/src/storage.rs:75:45
    |
75  |     env.storage().persistent().set(address, balance);
    |                                ---          ^^^^^^^ expected `&_`, found `i128`
    |                                |
    |                                arguments to this method are incorrect
    |
    = note: expected reference `&_`
                    found type `i128`
```